binary_search_tree/tree.py

```diff
@@ -9,40 +9,95 @@ def __init__(self, key, val = None):
+        
+        if new_node.key < current_root.key:
+            current_root.left = self.add_helper(current_root.left, new_node)
+        else:
+            current_root.right = self.add_helper(current_root.right, new_node)
+        return current_root
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if self.root is None:
+            self.root = new_node
+            return
+        self.add_helper(self.root, new_node)
+
+    def find_helper(self, current, key):
+        if current is None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif current.key <= key:
+            return self.find_helper(current.right, key)
+        return self.find_helper(current.left, key)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def find(self, key):
-        pass
+        if self.root is None:
+            return None
+        else:
+            return self.find_helper(self.root, key)
+
+    def inorder_helper(self, current, items):
+        if current is not None:
+            self.inorder_helper(current.left, items)
+            items.append({"key": current.key, "value": current.value})
+            self.inorder_helper(current.right, items)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper(self.root, items)
+        return items
+
+    def preorder_helper(self, current, items):
+        if current != None:
+            items.append({"key": current.key, "value": current.value})
+            self.preorder_helper(current.left, items)
+            self.preorder_helper(current.right, items)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
+
+    def postorder_helper(self, current, items):
+        if current != None:
+            self.postorder_helper(current.left, items)
+            self.postorder_helper(current.right, items)
+            items.append({"key": current.key, "value": current.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)    
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
+
+    def height_helper(self, current):
+        if current is None:
+            return 0
+        return 1 + max(self.height_helper(current.left), self.height_helper(current.right))
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def height(self):
```

Comment:
✨Nice! Time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!